roachtest: sqlsmith/setup=tpch-sf1/setting=no-mutations failed

[(roachtest).sqlsmith/setup=tpch-sf1/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2707822&tab=buildLog) on [master@5cfd7e5553a3072a1490d392390dddf968844215](https://github.com/cockroachdb/cockroach/commits/5cfd7e5553a3072a1490d392390dddf968844215):

```
																		(2572262315:::OID)
																)
																	AS tab_58 (col_121)
														),
													with_31 (col_122)
														AS (
															SELECT
																*
															FROM
																(
																	VALUES
																		('04:39:49.711166':::TIME),
																		(NULL),
																		('09:25:56.177598':::TIME)
																)
																	AS tab_59 (col_122)
														)
												SELECT
													'\x3b177c':::BYTES AS col_123
												LIMIT
													1:::INT8
											)
												AS col_124
										FROM
											with_28 AS cte_ref_10
										ORDER BY
											cte_ref_10.col_118 DESC
										LIMIT
											1:::INT8
									)
								),
								('\x47ea':::BYTES),
								(NULL),
								('\x':::BYTES)
						)
							AS tab_60 (col_125)
				)
		SELECT
			tab_61.r_name AS col_126,
			tab_61.r_regionkey AS col_127,
			tab_61.r_comment AS col_128,
			tab_61.tableoid AS col_129,
			'\x':::BYTES AS col_130,
			NULL AS col_131
		FROM
			defaultdb.public.region@primary AS tab_61, with_32 AS cte_ref_11
		WHERE
			e'-\x1e\x02':::STRING NOT ILIKE default_to_database_primary_region(e'\x046\\\x04>9}':::STRING::STRING)::STRING
		ORDER BY
			tab_61.tableoid ASC;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=tpch-sf1/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2707822&tab=artifacts#/sqlsmith/setup=tpch-sf1/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dtpch-sf1%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

```
07:39:26 test.go:344: test failure: 	sqlsmith.go:198,sqlsmith.go:228,test_runner.go:767: error: pq: internal error: default_to_database_primary_region(): default_to_database_primary_region: UpdateDeadlineMaybe() called on leaf txn
```
 
Multiregion team - do you know what's up here?

--- a/pkg/ccl/importccl/import_table_creation.go
+++ b/pkg/ccl/importccl/import_table_creation.go
@@ -208,6 +208,14 @@ func (so *importSequenceOperators) GetSerialSequenceNameFromColumn(
 	return nil, errors.WithStack(errSequenceOperators)
 }
 
+// CurrentDatabaseRegionConfig is part of the tree.EvalDatabase interface.
+func (so *importSequenceOperators) CurrentDatabaseRegionConfig() (
+	tree.DatabaseRegionConfig,
+	error,
+) {
+	return nil, errors.WithStack(errSequenceOperators)
+}
+
 // Implements the tree.EvalDatabase interface.
 func (so *importSequenceOperators) ParseQualifiedTableName(sql string) (*tree.TableName, error) {
 	name, err := parser.ParseTableName(sql)

--- a/pkg/sql/catalog/descpb/region_name.go
+++ b/pkg/sql/catalog/descpb/region_name.go
@@ -24,3 +24,24 @@ func (regions RegionNames) ToStrings() []string {
 	}
 	return ret
 }
+
+// IsValidRegionNameString implements the tree.DatabaseRegionConfig interface.
+func (cfg *DatabaseDescriptor_RegionConfig) IsValidRegionNameString(r string) bool {
+	if cfg == nil {
+		return false
+	}
+	for _, region := range cfg.Regions {
+		if string(region.Name) == r {
+			return true
+		}
+	}
+	return false
+}
+
+// PrimaryRegionString implements the tree.DatabaseRegionConfig interface.
+func (cfg *DatabaseDescriptor_RegionConfig) PrimaryRegionString() string {
+	if cfg == nil {
+		return ""
+	}
+	return string(cfg.PrimaryRegion)
+}

--- a/pkg/sql/faketreeeval/evalctx.go
+++ b/pkg/sql/faketreeeval/evalctx.go
@@ -44,6 +44,11 @@ func (so *DummySequenceOperators) GetSerialSequenceNameFromColumn(
 	return nil, errors.WithStack(errSequenceOperators)
 }
 
+// CurrentDatabaseRegionConfig is part of the tree.EvalDatabase interface.
+func (so *DummySequenceOperators) CurrentDatabaseRegionConfig() (tree.DatabaseRegionConfig, error) {
+	return nil, errors.WithStack(errSequenceOperators)
+}
+
 // ParseQualifiedTableName is part of the tree.EvalDatabase interface.
 func (so *DummySequenceOperators) ParseQualifiedTableName(sql string) (*tree.TableName, error) {
 	return nil, errors.WithStack(errSequenceOperators)
@@ -164,6 +169,11 @@ var _ tree.EvalPlanner = &DummyEvalPlanner{}
 
 var errEvalPlanner = pgerror.New(pgcode.ScalarOperationCannotRunWithoutFullSessionContext,
 	"cannot evaluate scalar expressions using table lookups in this context")
+
+// CurrentDatabaseRegionConfig is part of the tree.EvalDatabase interface.
+func (ep *DummyEvalPlanner) CurrentDatabaseRegionConfig() (tree.DatabaseRegionConfig, error) {
+	return nil, errors.WithStack(errEvalPlanner)
+}
 
 // ParseQualifiedTableName is part of the tree.EvalDatabase interface.
 func (ep *DummyEvalPlanner) ParseQualifiedTableName(sql string) (*tree.TableName, error) {

--- a/pkg/sql/region_util.go
+++ b/pkg/sql/region_util.go
@@ -772,3 +772,17 @@ func partitionByForRegionalByRow(
 		List:   listPartition,
 	}
 }
+
+// CurrentDatabaseRegionConfig is part of the tree.EvalDatabase interface.
+func (p *planner) CurrentDatabaseRegionConfig() (tree.DatabaseRegionConfig, error) {
+	_, dbDesc, err := p.Descriptors().GetImmutableDatabaseByName(
+		p.EvalContext().Ctx(),
+		p.txn,
+		p.CurrentDatabase(),
+		tree.DatabaseLookupFlags{Required: true},
+	)
+	if err != nil {
+		return nil, err
+	}
+	return dbDesc.GetRegionConfig(), nil
+}

--- a/pkg/sql/sem/builtins/builtins.go
+++ b/pkg/sql/sem/builtins/builtins.go
@@ -4927,40 +4927,22 @@ the locality flag on node startup. Returns an error if no region is set.`,
 		tree.FunctionProperties{Category: categoryMultiRegion},
 		stringOverload1(
 			func(evalCtx *tree.EvalContext, s string) (tree.Datum, error) {
-				r, err := evalCtx.InternalExecutor.QueryRow(
-					evalCtx.Ctx(),
-					DefaultToDatabasePrimaryRegionBuiltinName,
-					evalCtx.Txn,
-					`SELECT regions @> array[$1::string], primary_region
-				FROM crdb_internal.databases WHERE name = $2`,
-					s,
-					evalCtx.SessionData.Database,
-				)
+				regionConfig, err := evalCtx.Sequence.CurrentDatabaseRegionConfig()
 				if err != nil {
 					return nil, err
 				}
-				if len(r) == 0 {
+				if regionConfig.IsValidRegionNameString(s) {
+					return tree.NewDString(s), nil
+				}
+				primaryRegion := regionConfig.PrimaryRegionString()
+				if primaryRegion == "" {
 					return nil, pgerror.Newf(
 						pgcode.InvalidDatabaseDefinition,
-						"current database %s does not exist",
+						"current database %s is not multi-region enabled",
 						evalCtx.SessionData.Database,
 					)
 				}
-				// If region has been added to the database.
-				if *(r[0].(*tree.DBool)) {
-					return tree.NewDString(s), nil
-				}
-				// Otherwise, return the primary region if it exists.
-				if r[1] != tree.DNull {
-					return r[1], nil
-				}
-
-				// Not seeing any rows means the database is not multi-region enabled.
-				return nil, pgerror.Newf(
-					pgcode.InvalidDatabaseDefinition,
-					"current database %s is not multi-region enabled",
-					evalCtx.SessionData.Database,
-				)
+				return tree.NewDString(primaryRegion), nil
 			},
 			types.String,
 			`Returns the given region if the region has been added to the current database.

--- a/pkg/sql/sem/tree/eval.go
+++ b/pkg/sql/sem/tree/eval.go
@@ -3003,9 +3003,20 @@ func (e *MultipleResultsError) Error() string {
 	return fmt.Sprintf("%s: unexpected multiple results", e.SQL)
 }
 
+// DatabaseRegionConfig is a wrapper around DatabaseDescriptor_RegionConfig
+// related methods which avoids a circular dependency between descpb and tree.
+type DatabaseRegionConfig interface {
+	IsValidRegionNameString(r string) bool
+	PrimaryRegionString() string
+}
+
 // EvalDatabase consists of functions that reference the session database
 // and is to be used from EvalContext.
 type EvalDatabase interface {
+	// CurrentDatabaseRegionConfig returns the RegionConfig of the current
+	// session database.
+	CurrentDatabaseRegionConfig() (DatabaseRegionConfig, error)
+
 	// ParseQualifiedTableName parses a SQL string of the form
 	// `[ database_name . ] [ schema_name . ] table_name`.
 	// NB: this is deprecated! Use parser.ParseQualifiedTableName when possible.

